### DISCUSSION TITLE
How to find the location of openblas when compiling mxnet from source?

### DISCUSSION THREAD

**Discussion by KexinFeng** [apache/incubator-mxnet#20335] [General]
Hi everyone,

I got stuck on this problem when compiling mxnet from source: it couldn't find OpenBLAS include, OpenBLAS lib and OpenBLAS. I've tried this solution under similar question: https://github.com/apache/incubator-mxnet/discussions/16587#discussioncomment-62992 by setting the environment variable:

OpenBLAS_HOME=                  C:\utils\OpenBLAS
OpenBLAS_INCLUDE_DIR=           C:\utils\OpenBLAS\include\openblas

Then when I compile mxnet from source in the project directory:
`python ci/build_windows.py -f WIN_CPU
`

I got error:
2021-06-05 00:10:08,877 MXNet Windows build helper
2021-06-05 00:10:08,889 Build flavour: WIN_CPU
2021-06-05 00:10:08,890 Detected Windows platform
2021-06-05 00:10:08,891 Using vcvars environment:
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat
2021-06-05 00:10:08,891 Found MXNet root: C:\Users\fenkexin\Desktop\mxnet
2021-06-05 00:10:08,907 Generating project with CMake:
"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat" && cmake -GNinja -DCMAKE_C_COMPILER=cl -DCMAKE_CXX_COMPILER=cl -DUSE_CUDA=OFF -DUSE_CUDNN=OFF -DUSE_OPENCV=ON -DUSE_OPENMP=ON -DUSE_BLAS=open -DUSE_LAPACK=ON -DUSE_DIST_KVSTORE=OFF -DBUILD_CPP_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Release C:\Users\fenkexin\Desktop\mxnet
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.10.0
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
-- The C compiler identification is MSVC 19.29.30037.0
-- The CXX compiler identification is MSVC 19.29.30037.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_CROSSCOMPILING FALSE
-- CMAKE_HOST_SYSTEM_PROCESSOR AMD64
-- CMAKE_SYSTEM_PROCESSOR AMD64
-- CMAKE_SYSTEM_NAME Windows
-- CMake version '3.20.3' using generator 'Ninja'
-- The ASM_MASM compiler identification is MSVC
-- Found assembler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/ml64.exe
-- Could not find OpenBLAS include. Turning OpenBLAS_FOUND off
-- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off
CMake Error at cmake/Modules/FindOpenBLAS.cmake:82 (MESSAGE):
  Could not find OpenBLAS
Call Stack (most recent call first):
  cmake/ChooseBlas.cmake:42 (find_package)
  CMakeLists.txt:316 (include)

-- Configuring incomplete, errors occurred!
See also "C:/Users/fenkexin/Desktop/mxnet/build/CMakeFiles/CMakeOutput.log".
See also "C:/Users/fenkexin/Desktop/mxnet/build/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "ci/build_windows.py", line 290, in <module>
    sys.exit(main())
  File "ci/build_windows.py", line 278, in main
    windows_build(args)
  File "ci/build_windows.py", line 167, in windows_build
    check_call(cmd, shell=True, env=env)
  File "C:\Programs\Python\Python38\lib\subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat" && cmake -GNinja -DCMAKE_C_COMPILER=cl -DCMAKE_CXX_COMPILER=cl -DUSE_CUDA=OFF -DUSE_CUDNN=OFF -DUSE_OPENCV=ON -DUSE_OPENMP=ON -DUSE_BLAS=open -DUSE_LAPACK=ON -DUSE_DIST_KVSTORE=OFF -DBUILD_CPP_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Release C:\Users\fenkexin\Desktop\mxnet' returned non-zero exit status 1.

How should I let cmake find my openblas?

By the way, I don't know how much this affects the problem, but my OS is win10.

**Comment by TristonC** [comment]
Could you @KexinFeng  try to set OpenBLAS_HOME to the direction which contains the three folders(bin, include, lib) of openblas. For example, I set it as  C:\Lib\OpenBLAS-0.3.13-x64

**Reply by KexinFeng** [reply]
Thanks for the reply. Yes, I have set:
OpenBLAS_HOME= C:\utils\OpenBLAS
This directory contains three folders: lib, include, share.

I have switched to wsl though.

**Reply by TristonC** [reply]
@KexinFeng What is your progress so far?

**Reply by KexinFeng** [reply]
Thanks for checking in! I currently sort of bypass this issue and working in the Windows Subsystem for Linux now.
